In [9]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [38]:
# Import dataframe products.csv
path = r'C:\Users\lefev\Documents\Data Analytics Course\Instacart Basket Analysis' 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [71]:
# Import dataframe orders wrangeled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency Checks

In [74]:
# Check orders df infomration
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [76]:
# Create a test dataframe for mixed-type columns
df_test = pd.DataFrame()

In [78]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [80]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [82]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [84]:
# Fixing mixed type column "mix" to string
df_test['mix'] = df_test['mix'].astype('str')

In [86]:
# Finding missing values with products.csv
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [88]:
# Create a subset of the products dataframe containing only the values in question with product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [90]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [92]:
# Checking the products.csv shape to compare number of rows
df_prods.shape

(49693, 5)

In [94]:
# Create a new dataframe cleaned called df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [96]:
df_prods_clean.shape

(49677, 5)

In [98]:
# Check for duplicate rows in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [100]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [102]:
# Addressing the duplicates
df_prods_clean.shape

(49677, 5)

In [104]:
# Create a new dataframe without the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [106]:
df_prods_clean_no_dups.shape

(49672, 5)

# Start of Task

In [109]:
# Run the describe for df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

# Yes having a max of 2.3 seems odd for order time, as it can go till 24
# Also having a min of 0.000 order time seems odd too. 

In [112]:
# Check for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [113]:
# Double check for mixed-type in another way
def check_mixed_types(df):
    return df.apply(lambda x: x.apply(type).nunique()).to_dict()

In [116]:
mixed_types = check_mixed_types(df_ords)
for column, type_count in mixed_types.items():
    if type_count > 1:
        print(f"Column '{column}' has mixed types")

In [117]:
# Checking for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

# We have found 206209 missing values in the days_since_prior_order column. This might be explained by customers who never ordered more than twice.
# If a customer only ordered once, then there would be no reason to fill that column. 

## Check for duplicates

In [126]:
df_dups_ords = df_ords[df_ords.duplicated()]

In [132]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

In [140]:
# Export dataframes
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_ords.csv'), index=False)
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_clean.csv'), index=False)
